components/scream/src/physics/rrtmgp/rrtmgp_inputs_initializer.cpp

```diff
@@ -0,0 +1,179 @@
+#include "physics/rrtmgp/scream_rrtmgp_interface.hpp"
+#include "rrtmgp_test_utils.hpp"
+#include "mo_garand_atmos_io.h"
+#include "YAKL.h"
+
+#include <array>
+
+namespace scream {
+
+    void RRTMGPInputsInitializer::add_field (const field_type &f) {
+        const auto& id = f.get_header().get_identifier();
+        m_fields.emplace(id.name(),f);
+        m_fields_id.insert(id);
+    }
+
+    void RRTMGPInputsInitializer::add_field (const field_type &f, const field_type& f_ref,
+                                             const remapper_ptr_type& remapper) {
+        if (m_remapper) {
+            // Sanity check
+            EKAT_REQUIRE_MSG (m_remapper->get_src_grid()->name()==remapper->get_src_grid()->name(),
+              "Error! A remapper was already set in RRTMGPInputsInitializer, but its src grid differs from"
+              "       the grid of the input remapper of this call.\n");
+        } else {
+            m_remapper = remapper;
+            m_remapper->registration_begins();
+        }
+  
+        const auto& id = f.get_header().get_identifier();
+        const auto& id_ref = f_ref.get_header().get_identifier();
+  
+        // To the AD, we only expose the fact that we init f_ref...
+        m_fields_id.insert(id_ref);
+  
+        // ...but RRTMGP only knows how to init f...
+        m_fields.emplace(id.name(),f);
+  
+        // ...hence, we remap to f_ref.
+        m_remapper->register_field(f, f_ref);
+    }
+
+    // =========================================================================================
+    void RRTMGPInputsInitializer::initialize_fields () {
+        // Safety check: if we're asked to init anything at all,
+        // then we should have been asked to init 13 fields.
+        int count = 0;
+        count += m_fields.count("pmid");
+        count += m_fields.count("pint");
+        count += m_fields.count("tmid");
+        count += m_fields.count("tint");
+        count += m_fields.count("col_dry");
+        count += m_fields.count("gas_vmr");
+        count += m_fields.count("sfc_alb_dir");
+        count += m_fields.count("sfc_alb_dif");
+        count += m_fields.count("mu0");
+        count += m_fields.count("lwp");
+        count += m_fields.count("iwp");
+        count += m_fields.count("rel");
+        count += m_fields.count("rei");
+
+        if (count==0) {
+          return;
+        }
+  
+        EKAT_REQUIRE_MSG(
+            count==13,
+            "Error! RRTMGPInputsInitializer is expected to init\n"
+            "       pmid, pint, tmid, tint, col_dry, gas_vmr, sfc_alb_dir, sfc_alb_dif,\n"
+            "       mu0, lwp, iwp, rel, rei,\n"
+            "       but only " + std::to_string(count) + " of those have been found.\n"
+            "       Please, check the atmosphere processes you are using,\n"
+            "       and make sure they agree on who's initializing each field.\n"
+        );
+  
+        // Get device views
+        auto d_pmid = m_fields.at("pmid").get_view();
```

Comment:
For reference, if/when we switch to dealing with padded versions of these fields, there is an example of how IO deals with reading contiguous data into a padded array here: https://github.com/E3SM-Project/scream/blob/master/components/scream/src/share/io/scream_scorpio_interface.cpp#L158